### PR TITLE
SBP Review M3

### PR DESCRIPTION
**Pull request by ayushmishra2005**

### CODE REVIEW

File: c-pallets/cacher/src/lib.rs [hunk 1]
```diff
@@ -150,22 +159,35 @@ pub mod pallet {
...
+			// Can we make it more secure and add checks to verify biller information?
+			// For ex: We can match payment receiving account from `Bill` with `CacherInfo`.
+
+			// What is the use of `expiration_time` in `Bill`? Can we add a check for the same?
+
+			// What is the use of `file_hash` and `slice_hash`? I don't see its usage anywhere.
+
+			// We can remove `.iter()`.
+
+			for bill in &bills {
```
Comment (shakawwq) [inline c-pallets/cacher/src/lib.rs:187]: The cacher golang client obtains `to`, `file_hash`, `slice_hash` and `expiration_time` from the transaction record. Then use these values in your business logic.

**Review by shakawwq — COMMENTED** [review]
I have already replied or fixed all the suggestions or questions you raised above. Please take the time to check. If you have more suggestions, we are very willing to continue to improve them.

File: c-pallets/cacher/src/lib.rs [hunk 2]
```diff
@@ -150,22 +159,35 @@ pub mod pallet {
 		}
 
 		/// Pay to cachers for downloading files.
-		///	
+		///
 		/// Parameters:
 		/// - `bills`: list of bill.
 		#[pallet::weight(T::WeightInfo::pay(bills.len() as u32))]
 		pub fn pay(origin: OriginFor<T>, bills: Vec<Bill<AccountOf<T>, BalanceOf<T>, T::Hash>>) -> DispatchResult {
+			// Can we add validation for origin? This should not be allowed by everyone.
```
[the remark this answers is not in the document]
Comment (ayushmishra2005) [reply c-pallets/cacher/src/lib.rs:167]: There could be potential security issues due to following issues.

1. No Specific Origin: Either we specify any origin something like this https://github.com/paritytech/substrate/blob/416b0f50bba519146ec7ea45a67980b45cd658e7/frame/nfts/src/lib.rs#L114 or validate the origin.
2. `Vec` is unbounded by definition. We could use [BoundedVec](https://docs.rs/bounded-vec/latest/bounded_vec/struct.BoundedVec.html).

[the remark this answers is not in the document]
Comment (shakawwq) [reply c-pallets/cacher/src/lib.rs:167]: 1. Just like https://github.com/paritytech/substrate/blob/416b0f50bba519146ec7ea45a67980b45cd658e7/frame/nfts/src/lib.rs#L1726 . In the pay function, the bills is paid with the balance of the sender, so it is impossible to specify that the sender must be a specific account.
2. fixed: https://github.com/CESSProject/cess/pull/122/files#diff-f57d6fb814b266ce180bc630151d0278ed4fac0b1adfd59e42519de17f69268eR137

**Pull request closed without merging**